src/iop/colorequal.c

```diff
@@ -0,0 +1,1598 @@
+  schedule(simd:static) aligned(a_full, b_full, corrections, UV: 64)
+#endif
+  for(size_t k = 0; k < pixels; k++)
+  {
+    // For each correction factor, we re-express it as a[0] * U + a[1] * V + b
+    float uv[2] = { UV[2 * k + 0], UV[2 * k + 1] };
+    // corrections[4 * k + 0] = a_full[6 * k + 0] * uv[0] + a_full[6 * k + 1] * uv[1] + b_full[4 * k + 0];
+    corrections[4 * k + 1] = a_full[4 * k + 0] * uv[0] + a_full[4 * k + 1] * uv[1] + b_full[2 * k + 0];
+    corrections[4 * k + 2] = a_full[4 * k + 2] * uv[0] + a_full[4 * k + 3] * uv[1] + b_full[2 * k + 1];
+  }
+
+  dt_free_align(a_full);
+  dt_free_align(b_full);
+}
+
+/*
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const i, void *const o,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  dt_iop_colorequal_data_t *d = (dt_iop_colorequal_data_t *)piece->data;
+  blur_2D_bspline((float *)i, (float *)o, roi_out->width, roi_out->height, 4, d->chroma_size);
+}
+*/
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const i, void *const o,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  dt_iop_colorequal_data_t *d = (dt_iop_colorequal_data_t *)piece->data;
+
+  const int ch = piece->colors;
+  const float *const restrict in = (float*)i;
+  float *const restrict out = (float*)o;
+
+  const size_t npixels = (size_t)roi_out->width * roi_out->height;
+
+  // STEP 0: prepare the RGB <-> XYZ D65 matrices
+  // see colorbalancergb.c process() for the details, it's exactly the same
+  const struct dt_iop_order_iccprofile_info_t *const work_profile = dt_ioppr_get_pipe_current_profile_info(self, piece->pipe);
+  if(work_profile == NULL) return; // no point
+
+  dt_colormatrix_t input_matrix;
+  dt_colormatrix_t output_matrix;
+  dt_colormatrix_mul(input_matrix, XYZ_D50_to_D65_CAT16, work_profile->matrix_in);
+  dt_colormatrix_mul(output_matrix, work_profile->matrix_out, XYZ_D65_to_D50_CAT16);
+
+  float *const restrict UV = dt_alloc_align_float(npixels * 2);
+  float *const restrict corrections = dt_alloc_align_float(npixels * 4);
+  float *const restrict L = dt_alloc_align_float(npixels);
+
+  const float white = Y_to_dt_UCS_L_star(d->white_level);
+
+  // STEP 1: convert image from RGB to darktable UCS LUV
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(ch, npixels, in, out, UV, L, corrections, input_matrix, d, white)  \
+  schedule(simd:static) aligned(in, out, UV, L, corrections, input_matrix : 64)
+#endif
+  for(size_t k = 0; k < npixels; k++)
+  {
+    const float *const restrict pix_in = __builtin_assume_aligned(in + k * ch, 16);
+    float *const restrict uv = UV + k * 2;
+
+    // Convert to XYZ D65
+    dt_aligned_pixel_t XYZ_D65 = { 0.f };
+    dot_product(pix_in, input_matrix, XYZ_D65);
+    for_each_channel(c, aligned(XYZ_D65)) XYZ_D65[c] = fmaxf(XYZ_D65[c], 0.f);
```

Comment:
@TurboGit i did a huge amount of testing the last days. **All** artefacts i could observe are related to feeding some "NaN" data into the guiding filters, not completely sure if it's really NaNs or if some other "bad" data like -inf or alike could do this too. (Call these "bads" hereafter)

1. i replaced all blurring filters with gaussian blurs and suddenly we had clearly defined "spots". If we do any of the boxing blurs all bads get propagated over the rest of the filter, this results in a strong tendency to black blocks at lower right border.
2. The first group of "introducers" of those bads had been NaNs fired up from the interpolator.
3. Second group can be strong noise / sharp local gradients.

Could you check on your side if changing this to `fmaxf(XYZ_D65[c], 1.0e-7f)` fixes the issues? Here everything seems to fine now.

Reply:
It seems that the NaN is from the xyY conversion, to me the following also fix the issue and seems cleaner:

```
diff --git a/src/iop/colorequal.c b/src/iop/colorequal.c
index aa10081ce1..b60daa61b7 100644
--- a/src/iop/colorequal.c
+++ b/src/iop/colorequal.c
@@ -697,11 +697,16 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
     // Convert to XYZ D65
     dt_aligned_pixel_t XYZ_D65 = { 0.f };
     dot_product(pix_in, input_matrix, XYZ_D65);
-    for_each_channel(c, aligned(XYZ_D65)) XYZ_D65[c] = fmaxf(XYZ_D65[c], 0.f);
+    for_each_channel(c, aligned(XYZ_D65))
+      XYZ_D65[c] = fmaxf(XYZ_D65[c], 0.f);
 
     // Convert to dt UCS 22 UV and store UV
     dt_aligned_pixel_t xyY = { 0.f };
     dt_XYZ_to_xyY(XYZ_D65, xyY);
+
+    for_each_channel(c, aligned(XYZ_D65))
+      xyY[c] = dt_isnan(xyY[c]) ? 0.0f : xyY[c];
+
     xyY_to_dt_UCS_UV(xyY, uv);
     L[k] = Y_to_dt_UCS_L_star(xyY[2]);
   }
```

The only issue I see is that when enabling the guided filter, some "color" are leaking in some area (I would have thought that the default values should still be neutral?). See for example in the integration test the nails:

Without guided filter:

![image](https://github.com/darktable-org/darktable/assets/467069/c9bf492b-a98f-4d27-8666-c704aade5f81)

With guided filter (color leaking):

![image](https://github.com/darktable-org/darktable/assets/467069/382fbd61-4b65-4cee-bd4d-0b32c6eba1ea)

Adjusting the "chroma prefilter size" avoid the issue.

Reply:
> Could you check on your side if changing this to `fmaxf(XYZ_D65[c], 1.0e-7f)` fixes the issues? Here everything seems to fine now.

I see now that I haven't answered your question :) So yes, this fixes the issue on my side. But it looks a bit more fragile that my proposal above. Sure avoiding pure 0 in favor of 1.0e-7 fixes some NaN, but I would say that it is a "side" effect instead of checking NaN at the right place (where they are generated).

Reply:
BTW, this is due to the implementation of `dt_XYZ_to_xyY`. If vector `XYZ` is all zero we generate `NaN`.

```
static inline void dt_XYZ_to_xyY(const dt_aligned_pixel_t XYZ, dt_aligned_pixel_t xyY)
{
  const float sum = XYZ[0] + XYZ[1] + XYZ[2];
  xyY[0] = XYZ[0] / sum;
  xyY[1] = XYZ[1] / sum;
  xyY[2] = XYZ[1];
}
```

Reply:
http://www.brucelindbloom.com/index.html?Eqn_XYZ_to_xyY.html

Can we set x, y  to the chromaticity coordinates of your reference white as hinted in the doc above?

Reply:
> > Could you check on your side if changing this to `fmaxf(XYZ_D65[c], 1.0e-7f)` fixes the issues? Here everything seems to fine now.
> 
> I see now that I haven't answered your question :) So yes, this fixes the issue on my side. But it looks a bit more fragile that my proposal above. Sure avoiding pure 0 in favor of 1.0e-7 fixes some NaN, but I would say that it is a "side" effect instead of checking NaN at the right place (where they are generated).

Absolutely :-)

> BTW, this is due to the implementation of `dt_XYZ_to_xyY`. If vector `XYZ` is all zero we generate `NaN`.

Exactly, i was looking for such a "NaN generator" but couldn't spot it.



> Can we set x, y to the chromaticity coordinates of your reference white as hinted in the doc above?

Very good spotted :-)

Reply:
There is one more issue - not related to the above discussion - there seems to be some top/bottom issue in the `blur_2D_bspline`. Also we might want to use the kahan variants of the box blurring for better results. Or even use gaussians.

Reply:
`blur_2D_bspline` is misnamed, as it does a single iteration box mean, rather than applying a B-spline (like the 1-4-6-4-1 convolutions you see all over the codebase).  Iterated box mean approaches a Gaussian blur in the limit as the number of iterations increases, and as the number of iterations increases,  so does the reach of bad input values - after N iterations, each can potentially affect N*filter_width pixels in each dimension.

The box mean will be faster *per iteration* than a Gaussian blur (probably even for the Kahan variant), but I'm pretty sure Gaussian is faster than the eight iterations of `dt_box_mean` the code originally used.

Reply:
Plus the border check is just bad as it will take those data into account multiple times.

Reply:
> If vector `XYZ` is all zero we generate `NaN`.

Or if one of the components is NaN or inf ...

Reply:
I did some further checks and can confirm that we have 4-channel-zeros already after `dot_product(pix_in, input_matrix, XYZ_D65);` 

Checking for NaNs or all-zeros in `dt_XYZ_to_xyY()` would be pretty costly ...

Reply:
Good progess here. Using a const rough guess
```
static inline void dt_XYZ_to_xyY(const dt_aligned_pixel_t XYZ, dt_aligned_pixel_t xyY)
{
  const float sum = XYZ[0] + XYZ[1] + XYZ[2];
  const gboolean good = sum != 0.0f;
  // in case we can't calculate something meaningful
  // we fall back to something very dark but mimicking channels
  // by some rough guess
  xyY[0] = good ? XYZ[0] / sum : 2.0e-9f;
  xyY[1] = good ? XYZ[1] / sum : 1.0e-9f;
  xyY[2] = good ? XYZ[1]       : 1.3e-9f;
}
```
already leads to very good results. color leaking at totally black areas in the image is mostly gone :-)

Also i modified the whole code a) using gaussian blurs all over b) modified the sigmas according to roi scale c) less downscaling for the first guiding loop. Quite good now ... not a patch but whole zipped colorequal.c 
[colorequal.zip](https://github.com/darktable-org/darktable/files/13867622/colorequal.zip)

Reply:
Hum... with this last version I have this:

![image](https://github.com/darktable-org/darktable/assets/467069/df1f81c7-2b19-4828-8d01-d69ff455106c)

When using D&S strong sharpening and the coloreq guided filter.

Reply:
I see I also need to patch `dt_XYZ_to_xyY` !

Reply:
Ok, lot better to me too now. I think that the default for the guided filter should probably be:
3 / 5 / 50 / 6

In my testing it seems that this works better and eliminate most of the blending, so maybe a safer default.

Reply:
Arrgh - stupid me, but got it. It must of course be
```
static inline void dt_XYZ_to_xyY(const dt_aligned_pixel_t XYZ, dt_aligned_pixel_t xyY)
{
  const gboolean black = XYZ[0] == 0.0f && XYZ[1] == 0.0f && XYZ[2] == 0.0f;
  /* the calculation for black would fail with NaNs as result.
     According to http://www.brucelindbloom.com/index.html?Eqn_XYZ_to_xyY.html
     we would want the chromaticity coordinates of the reference white.
     As this is a costly operation requiring dev->white-balance date we use a rough guess
     of 2.0, 1.0, 1.5
  */
  const float sum = XYZ[0] + XYZ[1] + XYZ[2];
  xyY[0] = black ? 2.0f / 4.5f : XYZ[0] / sum;
  xyY[1] = black ? 1.0f / 4.5f : XYZ[1] / sum;
  xyY[2] = XYZ[1];
}
``` 

After this we don't have blue leakings with the guided filter any more!